# Exploring White House Visitor Data

In this project I will provide a brief analysis of 2015 [White House Visitor Access Records](https://obamawhitehouse.archives.gov/briefing-room/disclosures/visitor-records). I will focus on the date and time data to answer the following questions:

### How many visitors did the White house have per month?

### What is the earliest and latest time of appointments?

### What are the minimum and maximum length of appointments?

### Why is this information important and what else can be analyzed from this data?

#### To accomplish this I will do the following:

1. Explore the Data
2. Clean the Data
3. Analyze the Data
4. Make Conclusions from my analysis

### 1. Explore the Data

In [1]:
from csv import reader
opened_file = open('potus_2015.csv')
read_file = reader(opened_file)
potus = list(read_file)
potus_h = potus[0]
potus = potus[1:]

In [2]:
def explore_data(dataset, start, end, rows_and_columns = False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n')

    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))
        
print(potus_h)
print('\n')
explore_data(potus, 0, 5, True)

['name', 'appt_made_date', 'appt_start_date', 'appt_end_date', 'visitee_namelast', 'visitee_namefirst', 'meeting_room', 'description']


['Joshua T. Blanton', '2014-12-18T00:00:00', '1/6/15 9:30', '1/6/15 23:59', '', 'potus', 'west wing', 'JointService Military Honor Guard']


['Jack T. Gutting', '2014-12-18T00:00:00', '1/6/15 9:30', '1/6/15 23:59', '', 'potus', 'west wing', 'JointService Military Honor Guard']


['Bradley T. Guiles', '2014-12-18T00:00:00', '1/6/15 9:30', '1/6/15 23:59', '', 'potus', 'west wing', 'JointService Military Honor Guard']


['Loryn F. Grieb', '2014-12-18T00:00:00', '1/6/15 9:30', '1/6/15 23:59', '', 'potus', 'west wing', 'JointService Military Honor Guard']


['Travis D. Gordon', '2014-12-18T00:00:00', '1/6/15 9:30', '1/6/15 23:59', '', 'potus', 'west wing', 'JointService Military Honor Guard']


Number of rows: 47954
Number of columns: 8


### 2. Clean the Data

In [3]:
def process_time(start):
    if ':' not in start:
        start_2 = start + ' 12:00'
        start = start_2
    else:
        return start
    return start

for row in potus:
    start = row[2]
    start = process_time(start)
    row[2] = start

start = []
for row in potus:
    start.append(row[2])
    
print(start[:5])

['1/6/15 9:30', '1/6/15 9:30', '1/6/15 9:30', '1/6/15 9:30', '1/6/15 9:30']


In [4]:
def process_time(end):
    if ':' not in end:
        end_2 = end + ' 12:00'
        end = end_2
    else:
        return end
    return end

for row in potus:
    end = row[3]
    end = process_time(end)
    row[3] = end

end = []
for row in potus:
    end.append(row[3])
    
print(end[:5])

['1/6/15 23:59', '1/6/15 23:59', '1/6/15 23:59', '1/6/15 23:59', '1/6/15 23:59']


In [5]:
import datetime as dt

In [6]:
date_format = "%m/%d/%y %H:%M"
for row in potus:
    appt_start_date = row[2]
    appt_start_date = dt.datetime.strptime(appt_start_date, date_format)
    row[2] = appt_start_date
print(type(appt_start_date))

date_format = "%m/%d/%y %H:%M"
for row in potus:
    appt_end_date = row[3]
    appt_end_date = dt.datetime.strptime(appt_end_date, date_format)
    row[3] = appt_end_date
print(type(appt_end_date))

<class 'datetime.datetime'>
<class 'datetime.datetime'>


### 3. Analyze the Data

#### How many visitors did the White house have per month?

In [7]:
visitors_per_month = {}
for row in potus:
    appt_start_date_obj = row[2]
    appt_start_date_str = appt_start_date_obj.strftime("%B, %y")
    if appt_start_date_str not in visitors_per_month:
         visitors_per_month[appt_start_date_str] = 1
    else:
        visitors_per_month[appt_start_date_str] += 1
print(visitors_per_month)

{'January, 15': 1248, 'February, 15': 2165, 'March, 15': 2262, 'April, 15': 4996, 'May, 15': 3013, 'June, 15': 7743, 'July, 15': 2930, 'August, 15': 1350, 'September, 15': 4416, 'October, 15': 3669, 'November, 15': 1133, 'December, 15': 13029}


In [8]:
appt_times = []
for row in potus:
    appt_dt = row[2]
    app_t = appt_dt.time()
    appt_times.append(app_t)
print(appt_times[:3])

[datetime.time(9, 30), datetime.time(9, 30), datetime.time(9, 30)]


#### What is the earliest and latest time of appointments?

In [9]:
min_time = min(appt_times)
print(min_time)
max_time = max(appt_times)
print(max_time)

06:00:00
21:30:00


#### What are the minimum and maximum length of appointments?

In [10]:
appt_lengths = {}
for row in potus:
    appt_s = row[2]
    appt_e = row[3]
    l = appt_e - appt_s
    if l not in appt_lengths:
        appt_lengths[l] = 1
    else:
        appt_lengths[l] += 1

min_length = min(appt_lengths)
max_length = max(appt_lengths)

print(min_length)
print(max_length)

2:29:00
16 days, 12:59:00


### 4. Conclusion

After a brief analysis of 2015 [White House Visitor Access Records](https://obamawhitehouse.archives.gov/briefing-room/disclosures/visitor-records) I can see that White House visits vary in many ways. For one, the start times run trhoughout the day from 6:00am to 9:30am and can last for many hours, and sometimes days. 

#### How many visitors did the White house have per month?

The range of visitors per month was from 1,133 in November to 7,743 in June of 2015.

#### What is the earliest and latest time of appointments?

The earliest time of appointments was 6:00am, and the latest was 9:30pm in 2015.

#### What are the minimum and maximum length of appointments?

The minimun length of appountments 2 hours and 29 minutes, and the maximum length of appointments was 16 days, 12 hours and 59 minutes.

#### Why is this information important and what else can be analyzed from this data?

This information is very important for many reasons. For one, this data should be analyzed thuroughly for security puposes. Because these records provide information on the entry and exit of individuals, length and frequency of appointments, and variations throughout the year, it can advise security planning, hiring, assessment, management and forecasting.